src/main/java/org/commcare/formplayer/session/FormSession.java

```diff
@@ -105,9 +106,17 @@ public FormSession(SerializableFormSession session,
         restoreFactory.setPermitAggressiveSyncs(false);
 
         this.sandbox = restoreFactory.getSandbox();
-        this.formDef = FormDefStringSerializer.deserialize(session.getFormXml());
-        loadInstanceXml(formDef, session.getInstanceXml());
-        formDef.setSendCalloutHandler(formSendCalloutHandler);
+
+        if (session.getFormDefinition() != null) {
+            this.formDef = FormDefStringSerializer.deserialize(session.getFormDefinition().getSerializedFormDef());
```

Comment:
@gherceg Think we should handle code breaking due to deserialization here. There can be changes in the underlying serialization methods (`readExternal` and `writeExternal`) due to which deserializing already serialized form def will break.  

Separately curious if we should clear `form_definition` table when user perform 'clear data' ?

Reply:
With the changes in https://github.com/dimagi/formplayer/pull/1127 I think it's simpler though we might have to store the raw XML as well so that we can re-parse it. This would be required especially for SMS sessions where we don't have the XML in an app DB.

Reply:
Does SMS sessions send the raw xml in the request data ? I am seeing only 2 path ways for someone to enter into a form - 
1. (NewFormResponseFactory](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java#L56) which uses xml from the request bean itself
2. [Normal menu navigation](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/session/MenuSession.java#L327) which seems to be getting the raw `FormDef` from the DB already

Does the sms uses sessions use the first pathway here ?

Would a solution here be just catching the deserialization exception and then reserialise raw form xml from one of the 2 sources (DB or request data) and store it in the `FormDefinition` postgress table ?

Reply:
> I think the form XML is only sent in the first request for SMS forms

Curious if that's the case why doesn't it error [here](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/services/NewFormResponseFactory.java#L63) already ?

Reply:
That code is only called on the first request (when the session is started). After that it calls the 'answer', 'current', 'next' and 'get-instance' endpoints which require an existing session.

Reply:
Thanks for raising Shubham and the additional feedback Simon. Just so I understand how this fits in with the changes made so far in #1127, at the moment that removes fetching the raw xml from the local db even in the case of normal menu navigation, and solely relies on the serialized xml from postgres which is risky. So the two changes that need to be made there are:
- fetch raw xml from db if deserialization fails in the FormDefinitionService (related to [this thread](https://github.com/dimagi/formplayer/pull/1127#discussion_r870368438))
- save raw xml to db when opening sms sessions so the previous step can be successful for sms sessions as well

Reply:
See 87f7c5039d82ff4f6c2a91ddbd62e7a47762bd8f